### DISCUSSION TITLE
How to obtain dFdXs gradients

### DISCUSSION THREAD

**Discussion by albisb** [mdolab/dafoam#499] [General]
Hi all,

I am trying to print out the total derivatives dFdXs for designVar = FFD, but so far I have not been successful. I am trying the following way. Inside pyDAFoam.py, I got to the point where dFdXs is calculated (inside the function mapdXvTodFFD) and added a json export:

        dFdXs = self.mesh.getdXs()
        dFdXs = self.mapVector(dFdXs, self.allWallsGroup, self.designSurfacesGroup)
        dFdFFD = self.DVGeo.totalSensitivity(dFdXs, ptSetName=self.ptSetName, comm=self.comm)
        import json
        import os
        Info("Current working directory:")
        Info(os.getcwd())
        with open("dFdXs_output.json","w") as f:
            json.dump(dFdXs, f)
        Info("dFdXs_output.json saved to "+os.getcwd())
        return dFdFFD
        
However, this doesn't seem to be working. I traced back to the function where mapdXvToFFD is called, which is solveAdjoint and added printouts as early as in the first line of code to check where is no longer entering the mapdXvToFFD, but none of those printouts was added to the log, so I am wondering when the function solveAdjoint is used, or if there may be other reasons for no printouts?

I also saw that there is a function called writeSurfaceSensitivityMap doing the same, but I also cannot call it for the same reasons. 

About my runScript:
- I tried both runAdjoint and opt
- Using FFD: "designVar": {
        "shape": {"designVarType": "FFD"},
- Running in reverse mode: prob.setup(mode="rev")
- I tried activating the "writeSensMap" flag as well in daoptions: "writeSensMap": ["shape"]

Finally, I would like to ask how can we use the bin files of dFdWColoring.bin, how should we read them?

Thanks

**Comment by friedenhe** [comment]
I suspect the reason is that you used a v3 script, which does not support "writeSensMap". Please refer to this v2 script on how to use this feature: https://github.com/mdolab/dafoam/blob/0cf80ff50972d1679b4c967c52e650744aaf1a59/tests/runTests_DASimpleFoamAD.py#L46

Also, please feel free to create an issue to request the "writeSensMap" feature for DAFoam-v3.

**Reply by albisb** [reply]
You're right, the v2 script fixed the issue and now I can get the sensitivities.
Thanks!
